### PR TITLE
Add support for stitching pre- and post-norm images in acquisition order

### LINKED ISSUE #22: Add tiling process for images before and after normalization
**Is your feature request related to a problem? Please describe.**

For specific channels in a run, we'll need to display a tiled image in acquisition order before and after normalization.

**Describe the solution you'd like**

Acquisition order is defined by the FOV number. It's assumed that there are no more -scan-2 images in a run.

There are two options that can be implemented

1. Visualize the before and after in separate images, and combine afterwards
2. Visualize the before and after in the same image

In case 1, we can use a process similar to #17 (and I'll likely export shared functionality to a separate function). Annotation will be the FOV number instead of the channel name. Padding will work the same.

Case 2 will be very similar to case 1, except there will need to be ample border space separating the before and after tiles. Indexing will need to take this into account when adding the after normalization images, since those will be to the right of the before normalization images.

(INPUT NEEDED) let me know which option works best.

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Closes #22.

**How did you implement your changes**

Add a function that takes both the pre- and post-norm directories, as well as the desired channel to stitch, and generate both stitched images.

One addition is the use of a generic stitching function for a 3-D `xarray` with annotation labels contained in the 0th dimension. I will be adding this to #17 as well.

**Remaining issues**

It may be better for this function to support just one directory. The user can then make multiple calls for both the pre- and post-norm stitched.

### CODE REVIEW

**Comment by alex-l-kong** [comment]
For this review my main questions are:

1. Whether it's better to have the function take both the pre-norm and post-norm directories, or whether it should just take one and have the user make two separate calls.
2. The structure of the generalized stitching function.

File: python_files/generate_supplementary_plots.py [hunk 1]
```diff
@@ -13,6 +23,82 @@
 import seaborn as sns
 
 
+def stitch_and_annotate_padded_img(image_data: xr.DataArray, padding: int = 25,
+                                   font_size: int = 100):
+    """Stitch an image with (x, y, z) dimensions, and annotate each image with labels contained
```
Comment (srivarra) [inline python_files/generate_supplementary_plots.py]: Is the `x` dimension the channel `c`?

File: python_files/generate_supplementary_plots.py [hunk 2]
```diff
@@ -13,6 +23,82 @@
...
+            stitched_image[
+                (row * row_len + padding * row) : ((row + 1) * row_len + padding * row),
+                (col * col_len + padding * col) : ((col + 1) * col_len + padding * col)
+            ] = image_data[img_idx, ...]
+            img_idx += 1
+            if img_idx == len(annotation_labels):
+                break
+
+    # define a draw instance for annotating the channel name
+    stitched_image_im: Image = Image.fromarray(stitched_image)
```
Comment (srivarra) [inline python_files/generate_supplementary_plots.py]: How is PIL? Do you find it easy to use?

**Review by srivarra — CHANGES_REQUESTED** [review]
Looks good, just a comment on dimension clarification. 

> Whether it's better to have the function take both the pre-norm and post-norm directories, or whether it should just take one and have the user make two separate calls.

I think just one call makes it easy to use.

>The structure of the generalized stitching function.

`stitch_and_annotate_padded_img` looks good, but the dimensions should be `(c,y,x)` or `(c,x,y)`.

File: python_files/generate_supplementary_plots.py (hunk 1 above)
Comment (alex-l-kong) [reply python_files/generate_supplementary_plots.py]: Yes, I'll rename the dimension variables so its clear.

File: python_files/generate_supplementary_plots.py (hunk 2 above)
Comment (alex-l-kong) [reply python_files/generate_supplementary_plots.py]: Haven't used it too much, it's nice for basic image manipulation and analysis such as simple morphology and filtering. IMO though it's not the easiest for more complex analysis, especially conversions. I've found `opencv` easier for even tasks such as converting a grayscale to RGB.

File: python_files/generate_supplementary_plots.py [hunk 3]
```diff
@@ -26,8 +112,77 @@
...
+        specified_run=run_name,
+        all_pre_norm_runs=all_pre_runs
+    )
+    verify_in_list(
+        specified_run=run_name,
+        all_post_norm_runs=all_norm_runs
+    )
+
+    # verify save_dir is valid before defining the save paths
+    validate_paths([save_dir])
```
Comment (camisowers) [inline python_files/generate_supplementary_plots.py]: So we're assuming the save_dir already exists or this will fail right?

**Review by camisowers — CHANGES_REQUESTED** [review]
Functions look good other than Sri's comments! 

The only thing missing is the short code ran to actually generate the TONIC comparison images we'll be publishing (i.e. looping through all the intended channels and calling `stitch_before_after_norm()`. Since this repo is serving as reference code for the paper, it's good to hold the reader's hand as much as possible.

**Comment by alex-l-kong** [comment]
> Functions look good other than Sri's comments!
> 
> The only thing missing is the short code ran to actually generate the TONIC comparison images we'll be publishing (i.e. looping through all the intended channels and calling `stitch_before_after_norm()`. Since this repo is serving as reference code for the paper, it's good to hold the reader's hand as much as possible.

That's a good point, I'll add another function that explicitly loads in the TONIC data and calls it.

File: python_files/generate_supplementary_plots.py (hunk 3 above)
Comment (alex-l-kong) [reply python_files/generate_supplementary_plots.py]: Yes, it should exist.

File: python_files/generate_supplementary_plots.py [hunk 4]
```diff
@@ -1,17 +1,119 @@
...
 import seaborn as sns
 
+SUPPLEMENTARY_FIG_DIR = "/Volumes/Shared/Noah Greenwald/TONIC_Cohort/supplementary_figs"
+ACQUISITION_INDICES = [
+    11, 12, 13, 14, 15, 17, 18, 20, 22, 23, 24, 28, 29, 30, 31, 32, 33, 34, 35,
+    36, 39, 40, 41, 42, 43, 44, 45, 46, 47
+]
+
+
+def stitch_and_annotate_padded_img(image_data: xr.DataArray, padding: int = 25,
```
Comment (ngreenwald) [inline python_files/generate_supplementary_plots.py]: Will you create a separate file to hold functions, and move both these to that file? Then import them here

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, just one organizational change

**Review by ngreenwald — APPROVED** [review]

**Pull request merged**